File: reflex/ext/reflex/pointer.cpp
#include "reflex/ruby/pointer.h"


#include <rays/ruby/point.h>
#include "defs.h"


RUCY_DEFINE_VALUE_FROM_TO(Reflex::Pointer)

#define THIS  to<Reflex::Pointer*>(self)

#define CHECK RUCY_CHECK_OBJ(Reflex::Pointer, self)


static
RUCY_DEF_ALLOC(alloc, klass)
{
	return new_type<Reflex::Pointer>(klass);
}
RUCY_END

static
RUCY_DEF6(initialize, type, action, position, modifiers, click_count, drag)
{
	CHECK;

	*THIS = Reflex::Pointer(
		to<uint>(type),
		(Reflex::Pointer::Action) to<int>(action),
		to<Rays::Point>(position),
		to<uint>(modifiers),
		to<uint>(click_count),
		to<bool>(drag));
	return self;
}
RUCY_END

static
RUCY_DEF1(initialize_copy, obj)
{
	CHECK;
	*THIS = to<Reflex::Pointer&>(obj);
	return self;
}
RUCY_END

static
RUCY_DEF0(get_type)
{
	CHECK;
	return value(THIS->type());
}
RUCY_END

static
RUCY_DEF0(get_action)
{
	CHECK;
	return value(THIS->action());
}
RUCY_END

static
RUCY_DEF0(get_position)
{
	CHECK;
	return value(THIS->position());
}
RUCY_END

static
RUCY_DEF0(get_modifiers)
{
	CHECK;
	return value(THIS->modifiers());
}
RUCY_END

static
RUCY_DEF0(get_click_count)
{
	CHECK;
	return value(THIS->click_count());
}
RUCY_END

static
RUCY_DEF0(is_drag)
{
	CHECK;
	return value(THIS->is_drag());
}
RUCY_END


static Class cPointer;

void
Init_pointer ()
{
	Module mReflex = define_module("Reflex");

	cPointer = mReflex.define_class("Pointer");
	cPointer.define_alloc_func(alloc);
	cPointer.define_private_method("initialize",      initialize);
	cPointer.define_private_method("initialize_copy", initialize_copy);
	cPointer.define_private_method("get_type",   get_type);
	cPointer.define_private_method("get_action", get_action);
	cPointer.define_method("position",    get_position);
	cPointer.define_method("modifiers",   get_modifiers);
	cPointer.define_method("click_count", get_click_count);
	cPointer.define_method("drag?",       is_drag);
	cPointer.define_const("TYPE_NONE",    Reflex::Pointer::TYPE_NONE);
	cPointer.define_const("MOUSE_LEFT",   Reflex::Pointer::MOUSE_LEFT);
	cPointer.define_const("MOUSE_RIGHT",  Reflex::Pointer::MOUSE_RIGHT);
	cPointer.define_const("MOUSE_MIDDLE", Reflex::Pointer::MOUSE_MIDDLE);
	cPointer.define_const("TOUCH",        Reflex::Pointer::TOUCH);
	cPointer.define_const("PEN",          Reflex::Pointer::PEN);
	cPointer.define_const("ACTION_NONE", Reflex::Pointer::ACTION_NONE);
	cPointer.define_const("DOWN",        Reflex::Pointer::DOWN);
	cPointer.define_const("UP",          Reflex::Pointer::UP);
	cPointer.define_const("MOVE",        Reflex::Pointer::MOVE);
	cPointer.define_const("CANCEL",      Reflex::Pointer::CANCEL);
	cPointer.define_const("STAY",        Reflex::Pointer::STAY);
}


namespace Reflex
{


	Class
	pointer_class ()
	{
		return cPointer;
	}


}// Reflex

File: reflex/lib/reflex/pointer.rb
# -*- coding: utf-8 -*-


require 'xot/bit_flag_accessor'
require 'xot/const_symbol_accessor'
require 'reflex/ext'


module Reflex


  class Pointer

    include Comparable

    alias types  get_type
    alias action get_action

    bit_flag_reader :types, **{
      none:         TYPE_NONE,
      mouse:        MOUSE,
      mouse_left:   MOUSE_LEFT,
      mouse_right:  MOUSE_RIGHT,
      mouse_middle: MOUSE_MIDDLE,
      touch:        TOUCH,
      pen:          PEN
    }

    const_symbol_reader :action, **{
      none:   ACTION_NONE,
      down:   DOWN,
      up:     UP,
      move:   MOVE,
      cancel: CANCEL,
      stay:   STAY
    }

    def mouse?()
      (get_type & MOUSE) != 0
    end

    def mouse_left?()
      (get_type & MOUSE_LEFT) != 0
    end

    def mouse_right?()
      (get_type & MOUSE_RIGHT) != 0
    end

    def mouse_middle?()
      (get_type & MOUSE_MIDDLE) != 0
    end

    alias left?   mouse_left?
    alias right?  mouse_right?
    alias middle? mouse_middle?

    def touch?()
      (get_type & TOUCH) != 0
    end

    def pen?()
      (get_type & PEN) != 0
    end

    def down?()
      get_action == DOWN
    end

    def up?()
      get_action == UP
    end

    def move?()
      get_action == MOVE
    end

    def cancel?()
      get_action == CANCEL
    end

    def stay?()
      get_action == STAY
    end

    alias pos position

    def x()
      position.x
    end

    def y()
      position.y
    end

    def <=>(o)
      inspect <=> o.inspect
    end

    def inspect()
      "#<Reflex::Pointer id:#{id} #{types} #{action} (#{x.round 2}, #{y.round 2}) mod:#{modifiers} click:#{click_count} drag:#{drag?} time:#{time.round 2}>"
    end

  end# Pointer


end# Reflex
